rocksdb: Bump version to 4.9 and fix broken interface usage

a-robinson opened the pull request:
They changed a couple parts of their interfaces that broke our code.
For now, I've taken the hideous approach of copying the old logic to
avoid modifying the options that we end up using, but will look for a
better solution tomorrow (and have asked their maintainers for a
recommendation as well).

Rushing this a bit because we think that 4.9 might fix a deadlock in
rocksdb that's regularly affecting us on GCE, and we'd like to verify
that assumption tomorrow morning and get it into this week's beta
release if it helps things.

Depends on https://github.com/cockroachdb/c-rocksdb/pull/24. Motivated by https://github.com/facebook/rocksdb/issues/1297 and https://github.com/cockroachdb/cockroach/issues/8784

Successfully builds and passes unit tests. The acceptance tests are in the middle of running.

@tschottdorf @tamird @sploiselle @petermattis

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8815)

<!-- Reviewable:end -->

a-robinson committed:
rocksdb: Bump version to 4.9 and fix broken interface usage

They changed a couple parts of their interfaces that broke our code.
For now, I've taken the hideous approach of copying the old logic to
avoid modifying the options that we end up using, but will look for a
better solution tomorrow (and have asked their maintainers for a
recommendation as well on https://reviews.facebook.net/D59319).

Rushing this a bit because we think that 4.9 might fix a deadlock in
rocksdb that's regularly affecting us on GCE, and we'd like to verify
that assumption tomorrow morning and get it into this week's beta
release if it helps things.

a-robinson commented:
circleci/teamcity won't be able to pass until https://github.com/cockroachdb/c-rocksdb/pull/24 is merged.

Also, Sean, if you'd like to try using a binary built with this following the steps that you've been using, you can download the binary from https://storage.googleapis.com/cockroach-alex/cockrach

tschottdorf commented:
LGTM in spirit. The code changes look innocuous enough, and for anyone else following, @a-robinson spent some time looking for a replacement of that deprecated OptionsBuilder to no avail.

@a-robinson if there is an upstream issue in which you inquire about the proper solution to the hack, mind linking that into your TODO in `db.cc` (and the commit message, while you're at it)?

Some back story for posterity:

This bug puts us in an odd situation: @sploiselle manages to run into it in 100% of the cases he sets up a GCE cluster when running the old beta release, yet to the best of my knowledge we have not seen it elsewhere, at least not regularly. But there's nothing noteworthy about his setup, so this is a pretty fundamental deadlock.
We're currently testing our `master` build to verify it has the same issue. Assuming (and hoping) it does, we're relatively certain this update improves stability (though the RocksDB changelog is very sparse, and in particular doesn't mention the fixed deadlock at all - we could be incurring an unrelated bug by updating). But if `master` doesn't have the problem, we won't really be certain this fixes anything.

Of course we have every reason to fix this deadlock in the next beta release, so hopefully by EOD tomorrow we have confirmed that this solves that particular deadlock. I personally don't consider it likely that we are releasing on that same day since the other issue (now https://github.com/grpc/grpc-go/pull/859, formerly https://github.com/grpc/grpc-go/issues/846) hasn't merged at the time of writing, and we'll need at least a solid fraction of 24 hours to assert that all stability regressions from the last beta have been fixed, which makes even Friday a slightly ambitious release day.

cc @cockroachdb/stability and in particular @petermattis

tschottdorf commented:
Also, I think you're going to want to make this a merge against `master` - we could cherry-pick it over, but it's awkward.

tschottdorf commented:
For testing purposes, I think you could push this commit to a separate branch; I *think* glock should still be able to pick it up. That way, at least the sister PR on cockroachdb/cockroach can run its test and you can learn early if anything funky happens there (of course the tradeoff is that we have to remember to merge this PR first and then update the other with the actual SHA).

a-robinson commented:
Thanks for the suggestions. I've:

* Pushed to a new `v4.9` branch in the c-rocksdb repo that I'll clean up once we're done with this
* Redirected this PR to master and rebased as necessary
* Added a link to the place I asked for advice (https://reviews.facebook.net/D59319) to the TODO and commit message

andreimatei commented:
Nice find!

I think we should merge this in master and develop around the same time,
otherwise I think it'll be a pain to switch between the two since you'll
have to rebuild rocksdb every time (right?).

On Thu, Aug 25, 2016 at 12:31 AM, Alex Robinson <notifications@github.com>
wrote:

> Thanks for the suggestions. I've:
>
>    - Pushed to a new v4.9 branch in the c-rocksdb repo that I'll clean up
>    once we're done with this
>    - Redirected this PR to master and rebased as necessary
>    - Added a link to the place I asked for advice (
>    https://reviews.facebook.net/D59319
>    <https://reviews.facebook.net/D59319>) to the TODO and commit message
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8815#issuecomment-242278592>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcVKeozkL_mZjV-SN794sSulGy1f8ks5qjRqhgaJpZM4JsokR>
> .
>

tschottdorf commented:
Yes, I think following the merge to `master` there would be an immediate merge of `master` into `develop`.

a-robinson commented:
Updated to use the pushed commit hash from https://github.com/cockroachdb/c-rocksdb/pull/25

tschottdorf commented:
LGTM

a-robinson commented:
At Tamir's request, added @petermattis to the TODO and moved the copied code into its own files. I kept the Facebook copyright header in them -- is that how we should do this?

tamird commented:
LGTM. If the license header lint passes, I'm ok with it.

tschottdorf commented:
I'm not a lawyer, but at least we did the same thing when we cribbed something small from Docker the other day (`acceptance/namesgenerator.go`)

a-robinson commented:
Thanks, I'll merge if circleci passes. Acceptance tests also passed locally.
